<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="zh">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.13.2"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>source: Exception Handling</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<script type="text/javascript" src="clipboard.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
  $(function() { init_search(); });
/* @license-end */
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">source
   </div>
  </td>
    <td>        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <span id="MSearchSelect"                onmouseover="return searchBox.OnSearchSelectShow()"                onmouseout="return searchBox.OnSearchSelectHide()">&#160;</span>
          <input type="text" id="MSearchField" value="" placeholder="搜索" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.svg" alt=""/></a>
          </span>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- 制作者 Doxygen 1.13.2 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() { codefold.init(0); });
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function(){initNavTree('_exception_handling.html',''); initResizable(true); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">载入中...</div>
<div class="SRStatus" id="Searching">搜索中...</div>
<div class="SRStatus" id="NoMatches">未找到</div>
</div>
</div>
</div>
</div>

<div><div class="header">
  <div class="headertitle"><div class="title">Exception Handling</div></div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>目录</h3>
<ul>
  <li class="level1">
    <a href="#CatchAnExceptionFromARunningTaskflow">Catch an Exception from a Running Taskflow</a>
  </li>
  <li class="level1">
    <a href="#CatchAnExceptionFromASubflow">Catch an Exception from a Subflow</a>
  </li>
  <li class="level1">
    <a href="#CatchAnExceptionFromAnAsyncTask">Catch an Exception from an Async Task</a>
  </li>
  <li class="level1">
    <a href="#CatchAnExceptionFromACorunLoop">Catch an Exception from a Corun Loop</a>
  </li>
</ul>
</div>
<div class="textblock"><p>This chapters discusses how to handle exceptions from a submitted taskflow so you can properly catch or propagate exceptions in your workload.</p>
<h1><a class="anchor" id="CatchAnExceptionFromARunningTaskflow"></a>
Catch an Exception from a Running Taskflow</h1>
<p>When a task throws an exception, the executor will store that exception in the shared state referenced by the <a class="el" href="classtf_1_1_future.html" title="class to access the result of an execution">tf::Future</a> handle. You can catch that exception via calling the <code>get</code> method:</p>
<div class="fragment"><div class="line"><a class="code hl_class" href="classtf_1_1_executor.html">tf::Executor</a> <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>;</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_taskflow.html">tf::Taskflow</a> <a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>;</div>
<div class="line"> </div>
<div class="line"><a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>.emplace([](){ <span class="keywordflow">throw</span> std::runtime_error(<span class="stringliteral">&quot;exception&quot;</span>); });</div>
<div class="line"> </div>
<div class="line"><span class="keywordflow">try</span> {</div>
<div class="line">  <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>.run(<a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>).get();</div>
<div class="line">}</div>
<div class="line"><span class="keywordflow">catch</span>(<span class="keyword">const</span> std::runtime_error&amp; <a class="code hl_function" href="_cwise__array__power__array_8cpp.html#a9cb86ff854760bf91a154f2724a9101d">e</a>) {</div>
<div class="line">  std::cerr &lt;&lt; <a class="code hl_function" href="_cwise__array__power__array_8cpp.html#a9cb86ff854760bf91a154f2724a9101d">e</a>.what() &lt;&lt; std::endl;</div>
<div class="line">}</div>
<div class="ttc" id="a_cwise__array__power__array_8cpp_html_a9cb86ff854760bf91a154f2724a9101d"><div class="ttname"><a href="_cwise__array__power__array_8cpp.html#a9cb86ff854760bf91a154f2724a9101d">e</a></div><div class="ttdeci">Array&lt; double, 1, 3 &gt; e(1./3., 0.5, 2.)</div></div>
<div class="ttc" id="aclasstf_1_1_executor_html"><div class="ttname"><a href="classtf_1_1_executor.html">tf::Executor</a></div><div class="ttdoc">class to create an executor for running a taskflow graph</div><div class="ttdef"><b>定义</b> executor-dl.hpp:51</div></div>
<div class="ttc" id="aclasstf_1_1_taskflow_html"><div class="ttname"><a href="classtf_1_1_taskflow.html">tf::Taskflow</a></div><div class="ttdoc">class to create a taskflow object</div><div class="ttdef"><b>定义</b> taskflow.hpp:67</div></div>
<div class="ttc" id="apoisson_8hpp_html_aa56fe360bb0ae38e0082f49e394ff825"><div class="ttname"><a href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a></div><div class="ttdeci">void taskflow(int nx, int ny, double dx, double dy, double *f, int itold, int itnew, double *u, double *unew, int block_size, unsigned num_threads)</div><div class="ttdef"><b>定义</b> taskflow.cpp:6</div></div>
<div class="ttc" id="athread__pool_8cpp_html_a543e564a8407bbeac15cb2d929fec755"><div class="ttname"><a href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a></div><div class="ttdeci">tf::Executor executor</div><div class="ttdef"><b>定义</b> thread_pool.cpp:6</div></div>
</div><!-- fragment --><dl class="section attention"><dt>注意</dt><dd>As <a class="el" href="classtf_1_1_future.html" title="class to access the result of an execution">tf::Future</a> is derived from @std_future, it inherits all the exception handling behaviors defined by the C++ standard.</dd></dl>
<p>An exception will automatically cancel the execution of its parent taskflow. All the subsequent tasks that have dependencies on that exception task will not run. For instance, the following code defines two tasks, <code><a class="el" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a></code> and <code><a class="el" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a></code>, where <code><a class="el" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a></code> runs after <code><a class="el" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a></code>. When <code><a class="el" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a></code> throws an exception, the executor will cancel the execution of the taskflow, stopping every tasks that run after <code><a class="el" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a></code>. In this case, <code><a class="el" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a></code> will not run.</p>
<div class="fragment"><div class="line"><a class="code hl_class" href="classtf_1_1_executor.html">tf::Executor</a> <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>;</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_taskflow.html">tf::Taskflow</a> <a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>;</div>
<div class="line"> </div>
<div class="line"><a class="code hl_class" href="classtf_1_1_task.html">tf::Task</a> <a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a> = <a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>.emplace([](){ <span class="keywordflow">throw</span> std::runtime_error(<span class="stringliteral">&quot;exception on A&quot;</span>); });</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_task.html">tf::Task</a> <a class="code hl_typedef" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a> = <a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>.emplace([](){ std::cout &lt;&lt; <span class="stringliteral">&quot;Task B\n&quot;</span>; });</div>
<div class="line"><a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a>.precede(<a class="code hl_typedef" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a>);</div>
<div class="line"> </div>
<div class="line"><span class="keywordflow">try</span> {</div>
<div class="line">  <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>.run(<a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>).get();</div>
<div class="line">}</div>
<div class="line"><span class="keywordflow">catch</span>(<span class="keyword">const</span> std::runtime_error&amp; <a class="code hl_function" href="_cwise__array__power__array_8cpp.html#a9cb86ff854760bf91a154f2724a9101d">e</a>) {</div>
<div class="line">  std::cerr &lt;&lt; <a class="code hl_function" href="_cwise__array__power__array_8cpp.html#a9cb86ff854760bf91a154f2724a9101d">e</a>.what() &lt;&lt; std::endl;</div>
<div class="line">}</div>
<div class="ttc" id="abench__gemm_8cpp_html_a37a83060ac796961b44991c836f083f7"><div class="ttname"><a href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a></div><div class="ttdeci">Matrix&lt; SCALARB, Dynamic, Dynamic &gt; B</div><div class="ttdef"><b>定义</b> bench_gemm.cpp:36</div></div>
<div class="ttc" id="abench__gemm_8cpp_html_addc86e8508f14411ec98f521c520f875"><div class="ttname"><a href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a></div><div class="ttdeci">Matrix&lt; SCALARA, Dynamic, Dynamic &gt; A</div><div class="ttdef"><b>定义</b> bench_gemm.cpp:35</div></div>
<div class="ttc" id="aclasstf_1_1_task_html"><div class="ttname"><a href="classtf_1_1_task.html">tf::Task</a></div><div class="ttdoc">class to create a task handle over a node in a taskflow graph</div><div class="ttdef"><b>定义</b> task.hpp:150</div></div>
</div><!-- fragment --><div class="fragment"><div class="line">{.shell-session} </div>
<div class="line">~$ exception on <a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a></div>
<div class="line"><span class="preprocessor"># execution of taskflow is cancelled after an execution is thrown</span></div>
</div><!-- fragment --><p>When multiple tasks throw exceptions simultaneously, the executor will only catch one exception and store it in the shared state. Other exceptions will be silently ignored. For example, the following taskflow may concurrently throw two exceptions from task <code><a class="el" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a></code> and task <code>C</code>. Only one exception, either <code><a class="el" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a></code> or <code>C</code>, will be propagated.</p>
<div class="fragment"><div class="line"><a class="code hl_class" href="classtf_1_1_executor.html">tf::Executor</a> <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>;</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_taskflow.html">tf::Taskflow</a> <a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>;</div>
<div class="line"> </div>
<div class="line"><span class="keyword">auto</span> [<a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a>, <a class="code hl_typedef" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a>, <a class="code hl_define" href="test__buffer__node_8cpp.html#ac4cf4b2ab929bd23951a8676eeac086b">C</a>, <a class="code hl_variable" href="_eigen_solver___eigen_solver___matrix_type_8cpp.html#a925331f9b6e4844293d36a8df2256d38">D</a>] = <a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>.emplace(</div>
<div class="line">  []() { std::cout &lt;&lt; <span class="stringliteral">&quot;TaskA\n&quot;</span>; },</div>
<div class="line">  []() { </div>
<div class="line">    std::cout &lt;&lt; <span class="stringliteral">&quot;TaskB\n&quot;</span>;</div>
<div class="line">    <span class="keywordflow">throw</span> std::runtime_error(<span class="stringliteral">&quot;Exception on Task B&quot;</span>);</div>
<div class="line">  },</div>
<div class="line">  []() { </div>
<div class="line">    std::cout &lt;&lt; <span class="stringliteral">&quot;TaskC\n&quot;</span>; </div>
<div class="line">    <span class="keywordflow">throw</span> std::runtime_error(<span class="stringliteral">&quot;Exception on Task C&quot;</span>);</div>
<div class="line">  },</div>
<div class="line">  []() { std::cout &lt;&lt; <span class="stringliteral">&quot;TaskD will not be printed due to exception\n&quot;</span>; }</div>
<div class="line">);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a>.precede(<a class="code hl_typedef" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a>, <a class="code hl_define" href="test__buffer__node_8cpp.html#ac4cf4b2ab929bd23951a8676eeac086b">C</a>);  <span class="comment">// A runs before B and C</span></div>
<div class="line"><a class="code hl_variable" href="_eigen_solver___eigen_solver___matrix_type_8cpp.html#a925331f9b6e4844293d36a8df2256d38">D</a>.succeed(<a class="code hl_typedef" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a>, <a class="code hl_define" href="test__buffer__node_8cpp.html#ac4cf4b2ab929bd23951a8676eeac086b">C</a>);  <span class="comment">// D runs after  B and C</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordflow">try</span> {</div>
<div class="line">  <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>.<a class="code hl_function" href="classtf_1_1_executor.html#a519777f5783981d534e9e53b99712069">run</a>(<a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>).get();</div>
<div class="line">}</div>
<div class="line"><span class="keywordflow">catch</span>(<span class="keyword">const</span> std::runtime_error&amp; <a class="code hl_function" href="_cwise__array__power__array_8cpp.html#a9cb86ff854760bf91a154f2724a9101d">e</a>) {</div>
<div class="line">  <span class="comment">// caught either B&#39;s or C&#39;s exception</span></div>
<div class="line">  std::cout &lt;&lt; <a class="code hl_function" href="_cwise__array__power__array_8cpp.html#a9cb86ff854760bf91a154f2724a9101d">e</a>.what() &lt;&lt; std::endl;</div>
<div class="line">}</div>
<div class="ttc" id="a_eigen_solver___eigen_solver___matrix_type_8cpp_html_a925331f9b6e4844293d36a8df2256d38"><div class="ttname"><a href="_eigen_solver___eigen_solver___matrix_type_8cpp.html#a925331f9b6e4844293d36a8df2256d38">D</a></div><div class="ttdeci">MatrixXcd D</div><div class="ttdef"><b>定义</b> EigenSolver_EigenSolver_MatrixType.cpp:14</div></div>
<div class="ttc" id="aclasstf_1_1_executor_html_a519777f5783981d534e9e53b99712069"><div class="ttname"><a href="classtf_1_1_executor.html#a519777f5783981d534e9e53b99712069">tf::Executor::run</a></div><div class="ttdeci">tf::Future&lt; void &gt; run(Taskflow &amp;taskflow)</div><div class="ttdoc">runs a taskflow once</div><div class="ttdef"><b>定义</b> executor-dl.hpp:2067</div></div>
<div class="ttc" id="atest__buffer__node_8cpp_html_ac4cf4b2ab929bd23951a8676eeac086b"><div class="ttname"><a href="test__buffer__node_8cpp.html#ac4cf4b2ab929bd23951a8676eeac086b">C</a></div><div class="ttdeci">#define C</div><div class="ttdef"><b>定义</b> test_buffer_node.cpp:28</div></div>
</div><!-- fragment --><h1><a class="anchor" id="CatchAnExceptionFromASubflow"></a>
Catch an Exception from a Subflow</h1>
<p>When you join a subflow using <a class="el" href="classtf_1_1_subflow.html#a59fcac1323e70d920088dd37bd0be245" title="enables the subflow to join its parent task">tf::Subflow::join</a>, you can catch an exception thrown by its children tasks. For example, the following code catches an exception from the child task <code><a class="el" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a></code> of the subflow <code>sf</code>:</p>
<div class="fragment"><div class="line"><a class="code hl_class" href="classtf_1_1_executor.html">tf::Executor</a> <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>;</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_taskflow.html">tf::Taskflow</a> <a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>;</div>
<div class="line"> </div>
<div class="line"><a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>.emplace([](<a class="code hl_class" href="classtf_1_1_subflow.html">tf::Subflow</a>&amp; sf) {</div>
<div class="line">  <a class="code hl_class" href="classtf_1_1_task.html">tf::Task</a> <a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a> = sf.emplace([]() { </div>
<div class="line">    std::cout &lt;&lt; <span class="stringliteral">&quot;Task A\n&quot;</span>;</div>
<div class="line">    <span class="keywordflow">throw</span> std::runtime_error(<span class="stringliteral">&quot;exception on A&quot;</span>); </div>
<div class="line">  });</div>
<div class="line">  <a class="code hl_class" href="classtf_1_1_task.html">tf::Task</a> <a class="code hl_typedef" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a> = sf.emplace([]() { </div>
<div class="line">    std::cout &lt;&lt; <span class="stringliteral">&quot;Task B\n&quot;</span>; </div>
<div class="line">  });</div>
<div class="line">  <a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a>.precede(<a class="code hl_typedef" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a>);</div>
<div class="line">  </div>
<div class="line">  <span class="comment">// catch the exception</span></div>
<div class="line">  <span class="keywordflow">try</span> {</div>
<div class="line">    sf.join();</div>
<div class="line">  }</div>
<div class="line">  <span class="keywordflow">catch</span>(<span class="keyword">const</span> std::runtime_error&amp; re) {</div>
<div class="line">    std::cout &lt;&lt; <span class="stringliteral">&quot;exception thrown during subflow joining: &quot;</span> &lt;&lt; <a class="code hl_function" href="namespacectre.html#a1eb09093367d4fb6ecf857afc771533a">re</a>.what() &lt;&lt; <span class="charliteral">&#39;\n&#39;</span>;</div>
<div class="line">  }</div>
<div class="line">});</div>
<div class="line"> </div>
<div class="line"><a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>.<a class="code hl_function" href="classtf_1_1_executor.html#a519777f5783981d534e9e53b99712069">run</a>(<a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>).get();</div>
<div class="ttc" id="aclasstf_1_1_subflow_html"><div class="ttname"><a href="classtf_1_1_subflow.html">tf::Subflow</a></div><div class="ttdoc">class to construct a subflow graph from the execution of a dynamic task</div><div class="ttdef"><b>定义</b> flow_builder.hpp:1409</div></div>
<div class="ttc" id="anamespacectre_html_a1eb09093367d4fb6ecf857afc771533a"><div class="ttname"><a href="namespacectre.html#a1eb09093367d4fb6ecf857afc771533a">ctre::re</a></div><div class="ttdeci">CTRE_FLATTEN constexpr CTRE_FORCE_INLINE auto re() noexcept</div><div class="ttdef"><b>定义</b> ctre.hpp:5957</div></div>
</div><!-- fragment --><p>When an exception is thrown, it will cancel the execution of the parent subflow. All the subsequent tasks that depend on that exception task will not run. The above code example has the following output:</p>
<div class="fragment"><div class="line">{.shell-session} </div>
<div class="line">Task <a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a></div>
<div class="line">exception thrown during subflow joining: exception on <a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a></div>
</div><!-- fragment --><p>Uncaught exception will be propagated to the parent level until being explicitly caught. For example, the code below will propagate the exception to the parent of the subflow, which in this case in its taskflow.</p>
<div class="fragment"><div class="line"><a class="code hl_class" href="classtf_1_1_executor.html">tf::Executor</a> <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>;</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_taskflow.html">tf::Taskflow</a> <a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>;</div>
<div class="line"> </div>
<div class="line"><a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>.emplace([](<a class="code hl_class" href="classtf_1_1_subflow.html">tf::Subflow</a>&amp; sf) {</div>
<div class="line">  <a class="code hl_class" href="classtf_1_1_task.html">tf::Task</a> <a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a> = sf.emplace([]() {</div>
<div class="line">    std::cout &lt;&lt; <span class="stringliteral">&quot;Task A\n&quot;</span>;</div>
<div class="line">    <span class="keywordflow">throw</span> std::runtime_error(<span class="stringliteral">&quot;exception on A&quot;</span>);</div>
<div class="line">  });</div>
<div class="line">  <a class="code hl_class" href="classtf_1_1_task.html">tf::Task</a> <a class="code hl_typedef" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a> = sf.emplace([]() {</div>
<div class="line">    std::cout &lt;&lt; <span class="stringliteral">&quot;Task B\n&quot;</span>; </div>
<div class="line">  }); </div>
<div class="line">  <a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a>.precede(<a class="code hl_typedef" href="bench__gemm_8cpp.html#a37a83060ac796961b44991c836f083f7">B</a>);</div>
<div class="line"> </div>
<div class="line">  <span class="comment">// uncaught exception will propagate to the parent</span></div>
<div class="line">  sf.join();</div>
<div class="line">});</div>
<div class="line"> </div>
<div class="line"><span class="keywordflow">try</span></div>
<div class="line">{</div>
<div class="line">  <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>.<a class="code hl_function" href="classtf_1_1_executor.html#a519777f5783981d534e9e53b99712069">run</a>(<a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>).get();</div>
<div class="line">}</div>
<div class="line"><span class="keywordflow">catch</span> (<span class="keyword">const</span> std::runtime_error&amp; re)</div>
<div class="line">{</div>
<div class="line">  std::cout &lt;&lt; <span class="stringliteral">&quot;exception thrown from running the taskflow: &quot;</span> &lt;&lt; <a class="code hl_function" href="namespacectre.html#a1eb09093367d4fb6ecf857afc771533a">re</a>.what() &lt;&lt; <span class="charliteral">&#39;\n&#39;</span>;</div>
<div class="line">}</div>
</div><!-- fragment --><div class="fragment"><div class="line">{.shell-session} </div>
<div class="line">Task <a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a></div>
<div class="line">exception thrown from <a class="code hl_variable" href="empty_8c.html#af1f449cc09f8d36befcce07bc38c29c0">running</a> the <a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>: exception on <a class="code hl_typedef" href="bench__gemm_8cpp.html#addc86e8508f14411ec98f521c520f875">A</a></div>
<div class="ttc" id="aempty_8c_html_af1f449cc09f8d36befcce07bc38c29c0"><div class="ttname"><a href="empty_8c.html#af1f449cc09f8d36befcce07bc38c29c0">running</a></div><div class="ttdeci">static volatile int running</div><div class="ttdef"><b>定义</b> empty.c:41</div></div>
</div><!-- fragment --><h1><a class="anchor" id="CatchAnExceptionFromAnAsyncTask"></a>
Catch an Exception from an Async Task</h1>
<p>Similar to @std_future, <a class="el" href="classtf_1_1_executor.html#af960048056f7c6b5bc71f4f526f05df7" title="creates a parameterized asynchronous task to run the given function">tf::Executor::async</a> will store the exception in the shared state referenced by the returned @std_future handle.</p>
<div class="fragment"><div class="line"><a class="code hl_class" href="classtf_1_1_executor.html">tf::Executor</a> <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>;</div>
<div class="line"><span class="keyword">auto</span> fu = <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>.async([](){ <span class="keywordflow">throw</span> std::runtime_error(<span class="stringliteral">&quot;exception&quot;</span>); });</div>
<div class="line"><span class="keywordflow">try</span> {</div>
<div class="line">  fu.get();</div>
<div class="line">}</div>
<div class="line"><span class="keywordflow">catch</span>(<span class="keyword">const</span> std::runtime_error&amp; <a class="code hl_function" href="_cwise__array__power__array_8cpp.html#a9cb86ff854760bf91a154f2724a9101d">e</a>) {</div>
<div class="line">  std::cerr &lt;&lt; <a class="code hl_function" href="_cwise__array__power__array_8cpp.html#a9cb86ff854760bf91a154f2724a9101d">e</a>.what() &lt;&lt; std::endl;</div>
<div class="line">}</div>
</div><!-- fragment --><p>Running the program will show the exception message on the async task:</p>
<div class="fragment"><div class="line">{.shell-session} </div>
<div class="line">~$ exception</div>
</div><!-- fragment --><p>On the other hand, since <a class="el" href="classtf_1_1_executor.html#a0461cb2c459c9f9473c72af06af9c701" title="similar to tf::Executor::async but does not return a future object">tf::Executor::silent_async</a> does not return any future handle, any exception thrown from a silent-async task will be silently caught by the executor and (1) propagated to the its parent task if the parent task exists or (2) ignored if the parent task does not exist.</p>
<div class="fragment"><div class="line"><a class="code hl_class" href="classtf_1_1_taskflow.html">tf::Taskflow</a> <a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>;</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_executor.html">tf::Executor</a> <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>;</div>
<div class="line"> </div>
<div class="line"><span class="comment">// exception will be silently ignored</span></div>
<div class="line"><a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>.silent_async([](){ <span class="keywordflow">throw</span> std::runtime_error(<span class="stringliteral">&quot;exception&quot;</span>); });</div>
<div class="line"> </div>
<div class="line"><span class="comment">// exception will be propagated to the parent tf::Runtime task and then its Taskflow</span></div>
<div class="line"><a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>.emplace([&amp;](<a class="code hl_class" href="classtf_1_1_runtime.html">tf::Runtime</a>&amp; rt){</div>
<div class="line">  rt.<a class="code hl_function" href="classtf_1_1_runtime.html#a0ce29efa2106c8c5a1432e4a55ab2e05">silent_async</a>([](){ <span class="keywordflow">throw</span> std::runtime_error(<span class="stringliteral">&quot;exception&quot;</span>); });</div>
<div class="line">});</div>
<div class="line"><span class="keywordflow">try</span> {</div>
<div class="line">  <a class="code hl_function" href="poisson_8hpp.html#aa56fe360bb0ae38e0082f49e394ff825">taskflow</a>.get();</div>
<div class="line">}</div>
<div class="line"><span class="keywordflow">catch</span>(<span class="keyword">const</span> std::runtime_error&amp; re) {</div>
<div class="line">  std::cout &lt;&lt; <a class="code hl_function" href="namespacectre.html#a1eb09093367d4fb6ecf857afc771533a">re</a>.what() &lt;&lt; std::endl;</div>
<div class="line">}</div>
<div class="ttc" id="aclasstf_1_1_runtime_html"><div class="ttname"><a href="classtf_1_1_runtime.html">tf::Runtime</a></div><div class="ttdoc">class to include a runtime object in a task</div><div class="ttdef"><b>定义</b> runtime.hpp:35</div></div>
<div class="ttc" id="aclasstf_1_1_runtime_html_a0ce29efa2106c8c5a1432e4a55ab2e05"><div class="ttname"><a href="classtf_1_1_runtime.html#a0ce29efa2106c8c5a1432e4a55ab2e05">tf::Runtime::silent_async</a></div><div class="ttdeci">void silent_async(F &amp;&amp;f)</div><div class="ttdoc">runs the given function asynchronously without returning any future object</div><div class="ttdef"><b>定义</b> executor-dl.hpp:2450</div></div>
</div><!-- fragment --><h1><a class="anchor" id="CatchAnExceptionFromACorunLoop"></a>
Catch an Exception from a Corun Loop</h1>
<p>When you corun a graph via <a class="el" href="classtf_1_1_executor.html#a8fcd9e0557922bb8194999f0cd433ea8" title="runs a target graph and waits until it completes using an internal worker of this executor">tf::Executor::corun</a> or <a class="el" href="classtf_1_1_runtime.html#a1c772e90614302024cfa52fa86d75cac" title="co-runs the given target and waits until it completes">tf::Runtime::corun</a>, any exception will be thrown during the execution. For example, the code below will throw an exception during the execution of <code>taskflow1</code>:</p>
<div class="fragment"><div class="line"><a class="code hl_class" href="classtf_1_1_executor.html">tf::Executor</a> <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>;</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_taskflow.html">tf::Taskflow</a> taskflow1;</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_taskflow.html">tf::Taskflow</a> taskflow2;</div>
<div class="line"> </div>
<div class="line">taskflow1.<a class="code hl_function" href="classtf_1_1_flow_builder.html#a60d7a666cab71ecfa3010b2efb0d6b57">emplace</a>([](){</div>
<div class="line">  <span class="keywordflow">throw</span> std::runtime_error(<span class="stringliteral">&quot;exception&quot;</span>);</div>
<div class="line">}); </div>
<div class="line">taskflow2.<a class="code hl_function" href="classtf_1_1_flow_builder.html#a60d7a666cab71ecfa3010b2efb0d6b57">emplace</a>([&amp;](){</div>
<div class="line">  <span class="keywordflow">try</span> {</div>
<div class="line">    <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>.corun(taskflow1);</div>
<div class="line">  } <span class="keywordflow">catch</span>(<span class="keyword">const</span> std::runtime_error&amp; re) {</div>
<div class="line">    std::cout &lt;&lt; re.what() &lt;&lt; std::endl;</div>
<div class="line">  }</div>
<div class="line">}); </div>
<div class="line"><a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>.<a class="code hl_function" href="classtf_1_1_executor.html#a519777f5783981d534e9e53b99712069">run</a>(taskflow2).get();</div>
<div class="ttc" id="aclasstf_1_1_flow_builder_html_a60d7a666cab71ecfa3010b2efb0d6b57"><div class="ttname"><a href="classtf_1_1_flow_builder.html#a60d7a666cab71ecfa3010b2efb0d6b57">tf::FlowBuilder::emplace</a></div><div class="ttdeci">Task emplace(C &amp;&amp;callable)</div><div class="ttdoc">creates a static task</div><div class="ttdef"><b>定义</b> flow_builder.hpp:1261</div></div>
</div><!-- fragment --><p>We can observe the same behavior when using <a class="el" href="classtf_1_1_runtime.html#a1c772e90614302024cfa52fa86d75cac" title="co-runs the given target and waits until it completes">tf::Runtime::corun</a>:</p>
<div class="fragment"><div class="line"><a class="code hl_class" href="classtf_1_1_executor.html">tf::Executor</a> <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>;</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_taskflow.html">tf::Taskflow</a> taskflow1;</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_taskflow.html">tf::Taskflow</a> taskflow2;</div>
<div class="line"> </div>
<div class="line">taskflow1.<a class="code hl_function" href="classtf_1_1_flow_builder.html#a60d7a666cab71ecfa3010b2efb0d6b57">emplace</a>([](){</div>
<div class="line">  <span class="keywordflow">throw</span> std::runtime_error(<span class="stringliteral">&quot;exception&quot;</span>);</div>
<div class="line">}); </div>
<div class="line">taskflow2.<a class="code hl_function" href="classtf_1_1_flow_builder.html#a60d7a666cab71ecfa3010b2efb0d6b57">emplace</a>([&amp;](<a class="code hl_class" href="classtf_1_1_runtime.html">tf::Runtime</a>&amp; rt){</div>
<div class="line">  <span class="keywordflow">try</span> {</div>
<div class="line">    rt.<a class="code hl_function" href="classtf_1_1_runtime.html#a1c772e90614302024cfa52fa86d75cac">corun</a>(taskflow1);</div>
<div class="line">  } <span class="keywordflow">catch</span>(<span class="keyword">const</span> std::runtime_error&amp; re) {</div>
<div class="line">    std::cout &lt;&lt; re.what() &lt;&lt; std::endl;</div>
<div class="line">  }</div>
<div class="line">}); </div>
<div class="line"><a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>.<a class="code hl_function" href="classtf_1_1_executor.html#a519777f5783981d534e9e53b99712069">run</a>(taskflow2).get();</div>
<div class="ttc" id="aclasstf_1_1_runtime_html_a1c772e90614302024cfa52fa86d75cac"><div class="ttname"><a href="classtf_1_1_runtime.html#a1c772e90614302024cfa52fa86d75cac">tf::Runtime::corun</a></div><div class="ttdeci">void corun(T &amp;&amp;target)</div><div class="ttdoc">co-runs the given target and waits until it completes</div><div class="ttdef"><b>定义</b> executor-dl.hpp:2403</div></div>
</div><!-- fragment --><p>For the above example, if the exception is not caught with <a class="el" href="classtf_1_1_runtime.html#a1c772e90614302024cfa52fa86d75cac" title="co-runs the given target and waits until it completes">tf::Runtime::corun</a>, it will be propagated to its parent task, which is the <a class="el" href="classtf_1_1_runtime.html" title="class to include a runtime object in a task">tf::Runtime</a> object <code>rt</code> in this case. Then, the exception will be propagated to <code>taskflow2</code>.</p>
<div class="fragment"><div class="line"><a class="code hl_class" href="classtf_1_1_executor.html">tf::Executor</a> <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>;</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_taskflow.html">tf::Taskflow</a> taskflow1;</div>
<div class="line"><a class="code hl_class" href="classtf_1_1_taskflow.html">tf::Taskflow</a> taskflow2;</div>
<div class="line"> </div>
<div class="line">taskflow1.<a class="code hl_function" href="classtf_1_1_flow_builder.html#a60d7a666cab71ecfa3010b2efb0d6b57">emplace</a>([](){</div>
<div class="line">  <span class="keywordflow">throw</span> std::runtime_error(<span class="stringliteral">&quot;exception&quot;</span>);</div>
<div class="line">}); </div>
<div class="line">taskflow2.<a class="code hl_function" href="classtf_1_1_flow_builder.html#a60d7a666cab71ecfa3010b2efb0d6b57">emplace</a>([&amp;](<a class="code hl_class" href="classtf_1_1_runtime.html">tf::Runtime</a>&amp; rt){</div>
<div class="line">  rt.<a class="code hl_function" href="classtf_1_1_runtime.html#a1c772e90614302024cfa52fa86d75cac">corun</a>(taskflow1);</div>
<div class="line">}); </div>
<div class="line"> </div>
<div class="line"><span class="keywordflow">try</span> {</div>
<div class="line">  <a class="code hl_variable" href="thread__pool_8cpp.html#a543e564a8407bbeac15cb2d929fec755">executor</a>.run(taskflow2).get();</div>
<div class="line">}</div>
<div class="line"><span class="keywordflow">catch</span>(<span class="keyword">const</span> std::runtime_error&amp; re) {</div>
<div class="line">  std::cout &lt;&lt; <a class="code hl_function" href="namespacectre.html#a1eb09093367d4fb6ecf857afc771533a">re</a>.what() &lt;&lt; std::endl;</div>
<div class="line">}</div>
</div><!-- fragment --><p>For the above example, if the exception is not caught with <a class="el" href="classtf_1_1_runtime.html#a1c772e90614302024cfa52fa86d75cac" title="co-runs the given target and waits until it completes">tf::Runtime::corun</a>, it will be propagated to its parent task, which is the <a class="el" href="classtf_1_1_runtime.html" title="class to include a runtime object in a task">tf::Runtime</a> object <code>rt</code> in this case. Then, the exception will be propagated to <code>taskflow2</code>. </p>
</div></div><!-- contents -->
</div><!-- PageDoc -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">制作者 <a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.13.2 </li>
  </ul>
</div>
</body>
</html>
